File: Backend/php/modules/angular.php
<?php
if(!defined('prevent_direct_access')) { header('HTTP/1.0 403 Forbidden'); exit; }

class angular {
    public function RequiresAuthentication($context) {        
        if ($_SERVER['REQUEST_METHOD'] == 'GET') 
            return false;        
        return true;
    }

    public function Process($context) {
        $url = $_SERVER['REQUEST_URI'];
        if(endsWith($url, 'index.js'))
            return $this->IndexJS($context);
        else
            return $this->IndexHtml($context);
    }

    public function IndexJS($context) {        
        return "__CONFIG__ = Object.assign(__CONFIG__, {
            rest: '".scheme()."://".$_SERVER['HTTP_HOST']."'
            , auth: '".scheme()."://".$_SERVER['HTTP_HOST'].$context->nav['url']."'
            , angular_client: ".json_encode($context->nav)."
        })";
    }

    public function IndexHtml($context) {        
        // read index html
        $result = file_get_contents('wwwroot/index.tmpl');        

        //         
        $url = $_SERVER['REQUEST_URI'];
        if($url == '/') $url = '';
        
        $result = str_replace("@title", $context->nav['name'], $result);
        $result = str_replace("@base_href", "<base href='".$context->nav['url']."'>", $result);
        $result = str_replace("@path", $context->nav['url'], $result);            

        return $result;
    }

    public function Authenticated($context) {
        
        // get angular navigation
        $angular_navigation = array();
        
        $sql = "SELECT * FROM angular_navigation WHERE navigation_id=?";
        $result = $context->conn->query($sql, array($context->nav['_id']));
        if (count($result) > 0) {
            foreach($result as $ang_nav) {
                $content = json_decode($ang_nav['content'], true);                
                $content['_id'] = $ang_nav['_id'];
                $content['navigation_id'] = $context->nav['_id'];

                array_push($angular_navigation, $content);
            }
        }

        // get angular ui
        $angular_ui = array();
        
        // retrieve all matching uiElements and load it
        $uiElementIds = array();
        foreach($angular_navigation as $nav) {
            if(@$nav['uiElementIds']) $uiElementIds = array_merge($uiElementIds, $nav['uiElementIds']);
            if(@$nav['hidden'] && @$nav['hidden']['uiElementIds']) $uiElementIds = array_merge($uiElementIds, $nav['hidden']['uiElementIds']);
            if(@$nav['children']) {
                foreach($nav['children'] as $child) {
                    if(@$child['uiElementIds']) $uiElementIds = array_merge($uiElementIds, $child['uiElementIds']);
                    if(@$child['hidden'] && @$child['hidden']['uiElementIds']) $uiElementIds = array_merge($uiElementIds, $child['hidden']['uiElementIds']);
                }
            }
        }
        
        $sql = "SELECT * FROM angular_ui WHERE _id IN (".implode(", ", array_fill(0, count($uiElementIds), '?')).")";
        $result = $context->conn->query($sql, $uiElementIds);
        if (count($result) > 0) {
            foreach($result as $ang_ui) {                
                $angular_ui[$ang_ui['_id']] = json_decode($ang_ui['content']);
            }
        }

        return json_encode(array(
            "angular_navigation" => $angular_navigation
            , "angular_ui" => $angular_ui
        ));        
    }
}

?>

File: Backend/asp.netcore/Modules/Angular.cs
using Microsoft.AspNetCore.Http;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Web.Application.Lib;
using Web.Application.Services.DB;
using Web.Appliction.Lib;

namespace Web.Application.Modules
{
    public class Angular : IModule

    {
        public string Authenticated(HttpContext context)
        {
            // parameters
            var db = (SQL)context.Items["db"];
            var navigation = (IDictionary<string, object>)context.Items["navigation"];

            // get angular navigation
            var angular_navigation = new List<IDictionary<string, object>>();
            var param = new Dictionary<string, object>();
            param["navigation_id"] = navigation.Get("_id");
            var navigations = db.Query($@"SELECT * FROM angular_navigation WHERE navigation_id=@navigation_id", param);
            foreach (var nav in navigations)
            {
                var content = JsonConvert.DeserializeObject<IDictionary<string, object>>($"{nav["content"]}");
                content["_id"] = nav["_id"];
                content["navigation_id"] = $"{navigation.Get("_id")}";

                angular_navigation.Add(content);
            }

            // get angular ui
            IDictionary<string, object> angular_ui = new Dictionary<string, object>();

            // fetch angular ui
            // get all uiElementIds from angular_navigations
            JArray jsonAngNavs = JArray.Parse(
                JsonConvert.SerializeObject(angular_navigation)
            );

            // retrieve all matching uiElements and load it
            var uiElements = jsonAngNavs.SelectTokens("$..uiElementIds");
            List<string> ids = new List<string>();
            foreach (var uiElementIds in uiElements)
                foreach (var uiElementId in uiElementIds)
                    ids.Add($"{uiElementId}");

            var angular_uis = db.Query($@"SELECT * FROM angular_ui WHERE _id IN ({string.Join(",", ids.ToArray())})");
            foreach (var ui in angular_uis)
            {
                string content = $"{ui["content"]}";
                content = StringUtils.MultilineToSingle(content);
                angular_ui[$"{ui["_id"]}"] = JsonConvert.DeserializeObject(content);
            }

            return JsonConvert.SerializeObject(new
            {
                angular_navigation,
                angular_ui
            });
        }

        public Task<string> Process(HttpContext context)
        {
            string result = null;

            // parameters
            string WebRootPath = $"{context.Items["WebRootPath"]}";
            var navigation = (IDictionary<string, object>)context.Items["navigation"];

            // get relative path
            string navigationPath = navigation.Get("url")?.ToString();
            var regex = new Regex(Regex.Escape(navigationPath));
            string relativePath = regex.Replace($"{context.Request.Path}", "", 1);

            // check if the file exists in wwwroot
            relativePath = relativePath.Replace('/', Path.DirectorySeparatorChar);
            string filepath = Path.Combine(WebRootPath, relativePath);
            if (File.Exists(filepath))
                result = File.ReadAllText(filepath);
            
            // check if it's index.js
            else if (context.Request.Path.Value.Split("/").Last() == "index.js")
                result = IndexJS(context);
             
            // otherwise return index.html
            else
                result = IndexHtml(context);

            return Task.FromResult(result);
        }

        private string IndexJS(HttpContext context)
        {
            var navigation = (IDictionary<string, object>)context.Items["navigation"];
            var config = JsonConvert.DeserializeObject<IDictionary<string, object>>($"{navigation["content"]}");

            // calculate basePath
            config["rest"] = $"{context.Request.Scheme}://{context.Request.Host}{context.Request.PathBase}";
            config["auth"] = $"{context.Request.Scheme}://{context.Request.Host}{context.Request.PathBase}{navigation.Get("url")}";
            
            return $@"window.__CONFIG__ = {JsonConvert.SerializeObject(config)}";
        }

        private string IndexHtml(HttpContext context)
        {
            string result = string.Empty;

            // read index html
            string WebRootPath = $"{context.Items["WebRootPath"]}";
            string filepath = Path.Combine(WebRootPath, "index.html");
            result = File.ReadAllText(filepath);

            // get navigation info
            var navigation = (IDictionary<string, object>)context.Items["navigation"];

            // render html
            var basePath = $"{context.Request.PathBase}{navigation.Get("url")}";
            if (basePath == "/") basePath = "";
            
            result = result
                .Replace("@title", navigation.Get("name")?.ToString())
                .Replace("@base_href", $@"<base href='{basePath}'>")
                .Replace("@path", $@"{basePath}")
                ;
            return result;
        }
    }
}
